### PR TITLE
retry: bugfix: package requirements with git commits (#35057)

### PR DESCRIPTION
**Pull request by tgamblin**
Trying #35057 again, as it was broken on `develop`.  I guess the PR was not up to date, and it failed despite CI passing.  @scheibelp FYI.

- [x] Specs that define 'new' versions in the require: section need to generate associated facts to indicate that those versions are valid.
- [x] add test to verify success with unknown versions.

### CODE REVIEW

File: lib/spack/spack/solver/asp.py
Comment (haampie) [inline lib/spack/spack/solver/asp.py]: I didn't review the original pr, but this sounds like a bug to me. A git version is more like a specialized version, like pkg+x is more specialized than pkg. They shouldn't be equal?

**Comment by haampie** [comment]
To what extent is this pr a workaround for the fact that `@x` gets interpreted as specific version x versus any x dot something? My pr should solve that for the most part. Maybe good to coordinate those two?

**Comment by alalazo** [comment]
Failure probably related to https://github.com/spack/spack/pull/36258

**Comment by tgamblin** [comment]
@haampie: 
> To what extent is this pr a workaround for the fact that @x gets interpreted as specific version x versus any x dot something? My pr should solve that for the most part. Maybe good to coordinate those two?

AFAIK this is not a workaround for that at all.  I do not think that solves this problem.

File: lib/spack/spack/solver/asp.py
Comment (tgamblin) [inline lib/spack/spack/solver/asp.py]: @scheibelp: This should just ensure that all the CLI DeclaredVersions have a higher precedence than the others.  I think you needed the complex DeclaredVersion class above because you introduced a potentially duplicate `idx` here.

In general it's good for the weights in the concretizer (which the `idx` fields end up being used for) to be distinct, anyway, as it makes for deterministic answers.

File: lib/spack/spack/solver/asp.py
Comment (tgamblin) [inline lib/spack/spack/solver/asp.py]: @scheibelp: This should just ensure that all the CLI DeclaredVersions have a higher precedence than the others.  I think you needed the complex DeclaredVersion class above because you introduced a potentially duplicate `idx` here.

In general it's good for the weights in the concretizer (which the `idx` fields end up being used for) to be distinct, anyway, as it makes for deterministic answers.

**Comment by tgamblin** [comment]
@scheibelp: I think the main issue with this was the incompatibility with  #36258 that @alalazo mentioned.  I think that is all that needs to be fixed, but I also think the `DeclaredVersion` thing above can be simplified back to how it was if we just change the `idx` -- can you address those two things?

File: lib/spack/spack/solver/asp.py
Comment (tgamblin) [reply lib/spack/spack/solver/asp.py]: It's actually not possible to have a `Version` with the git syntax b/c there is no `Version` class -- it's a function. There is `VersionBase`, but VersionBase is not part of the public API:

```python
>>> from spack.version import ver
>>> ver("git.master-1.0")
GitVersion('git.master-1.0')
>>> spack.version.Version("git.master-1.0")
GitVersion('git.master-1.0')
```

```python
>>> from spack.version import *
>>> VersionBase("git.master=foo")
Traceback (most recent call last):
  File "<console>", line 1, in <module>
NameError: name 'VersionBase' is not defined
```

We *do* use `VersionBase` in a few places outside `version.py`, but not to construct anything; only to test if something is a version and to copy it with `Version`.  That is likely the result of some refactoring that could be cleaned up but it doesn't affect this.

The parser also will not produce a `VersionBase` with a git version in it.

So basically this test should be unnecessary -- we don't allow it.

Comment (scheibelp) [reply lib/spack/spack/solver/asp.py]: "bug" implies that things aren't working as intended, but part of https://github.com/spack/spack/pull/35057 was to establish and enforce what the intent *is* (e.g. `test_git_user_supplied_reference_satisfaction` was added in https://github.com/spack/spack/pull/35057)

if you have `hash=2.2` and `2.2`, then intuitively (and enforced by the code) they are equal: we are saying:

* "I have something equivalent to X"
* "I have X"

So I think it makes sense for them to be equal. based on other conversations, the satisfaction semantics need to be different. https://github.com/spack/spack/pull/35057 ensures that all of those intentions are enforced. If you disagree with the intent then that merits a conversation about the desired semantics.

File: lib/spack/spack/solver/asp.py
Comment (scheibelp) [reply lib/spack/spack/solver/asp.py]: For this to work, I would have to ensure that wherever a GitVersion and non-GitVersion would be compared, that their indices are different (so they would not actually be compared). We do not have a global index generator so for example cannot currently coordinate between DeclaredVersions introduced by `packages.yaml` requirements vs. preferences. I'm inclined to say that choosing a separate origin would be cleaner: e.g. `version_provenance.package_preferences` and `version_provenance.package_requirements` (if the origin just stays `version_provenance.packages_yaml` for both, then `add_concrete_versions_from_specs` method signature needs a new parameter to pass a base index to increment from).

Personally I think all of this is less clear than avoiding these problems in DeclaredVersion, and I'd be inclined to keep the updated definition regardless of whether we circumvent this particular problem (EDIT) with index/origin because otherwise the comparisons are fragile.

Comment (scheibelp) [reply lib/spack/spack/solver/asp.py]: Update: setting the origin different (and restoring the definition of DeclaredVersion) causes unit test failures (although it avoids symmetries in the solver). I'm pretty sure setting the indices differently will run into the same failure. I will update on that here in the next hour.

File: lib/spack/spack/solver/asp.py
Comment (haampie) [reply lib/spack/spack/solver/asp.py]: Equating those makes things hard to reason about and will definitely lead to subtle bugs. For example:

```python
>>> spack.version.Version("1.0") in set([spack.version.GitVersion("abcdef=1.0")])
False
>>> spack.version.Version("1.0") in [spack.version.GitVersion("abcdef=1.0")]
True
```

Arguably this is a bug already.

File: lib/spack/spack/solver/asp.py
Comment (haampie) [inline lib/spack/spack/solver/asp.py:1880]: I wouldn't use `==` here to start with, it's like expecting `1 == [1]` (in my versions PR I'm dropping "promotion" of 1 to [1] so it's more clear)

File: lib/spack/spack/version.py
Comment (haampie) [inline lib/spack/spack/version.py]:
```suggestion
```

File: lib/spack/spack/solver/asp.py
Comment (haampie) [reply lib/spack/spack/solver/asp.py]: Forgot to say that it breaks transitivity, as in `git_version_a == version && git_version_b == version` does not imply `git_version_a == git_version_b`. So, there's a congruence, there's no equality, it's a bug.

Comment (scheibelp) [reply lib/spack/spack/solver/asp.py]: Note [this](https://github.com/spack/spack/pull/36347#discussion_r1146707434) does not have to do with the discrepancy between `==` and `satisfies`, but rather with a failure to implement consistency between `==` and `.hash()`, you can see that if you replace

```
spack.version.Version("1.0") in [spack.version.GitVersion("abcdef=1.0")]
```

with 

```
spack.version.Version("1.0") in set([spack.version.GitVersion("abcdef=1.0")])
```

which will also return `False`. These edits change nothing about `==` for versions so this bug is in fact preexisting.

**Comment by haampie** [comment]
> AFAIK this is not a workaround for that at all.

You're right, it's mostly a workaround for `git_version == version` being true, which is a bug, as it breaks transitivity as well as the equality-implies-identical-hash contract.

File: lib/spack/spack/solver/asp.py
Comment (haampie) [reply lib/spack/spack/solver/asp.py]: Sure, it's pre-existing, but then imo it would be better to fix the bug instead of working around it in the `exact_match` related changes in this PR. And also the special-casing that warrants this long comment could be gone, which would also be an improvement in cognitive load on the next person that needs to understand this.

File: lib/spack/spack/solver/asp.py
Comment (scheibelp) [reply lib/spack/spack/solver/asp.py]: Note that this particular location is not the source of index-based equivalence since the versions are in fact not the same. That being said, I can change the index here if desired (it wouldn't be sufficient to address the index-based collisions that were occurring before, which is addressed by https://github.com/spack/spack/pull/36347/commits/c92938fe6afa6725388dd3e4c4234b793602675b)

File: lib/spack/spack/test/concretize_requirements.py
Comment (haampie) [inline lib/spack/spack/test/concretize_requirements.py]: Notice that you can't concat `file://` with a path in general (like windows), see `spack.util.url.path_to_file_url(...)`

**Comment by scheibelp** [comment]
* I have introduced distinct index/version-provenance everywhere to minimize user confusion about comparisons and eliminated some unnecessary checks in https://github.com/spack/spack/pull/36347/commits/c92938fe6afa6725388dd3e4c4234b793602675b, and that is sufficient to address all needs of this PR without using an extended definition of DeclaredVersion, BUT...
* DeclaredVersion is about what versions to declare in the solver, which means their comparison has more to do with `Version.satisfies` than with `Version.__eq__`
  * For example in `pkg_version_rules` it constructs a set with `sorted(set(self.declared_versions[pkg.name]), key=key_fn)`
* This PR respects (and enforces) all described meanings of (and differences between) `__eq__` and `.satisfies`
* Furthermore, the fact that there is no global index generator is a source of fragility that could easily reintroduce bugs into this code (hence I recommend keeping the extended definition of DeclaredVersion)

File: lib/spack/spack/solver/asp.py
Comment (scheibelp) [reply lib/spack/spack/solver/asp.py]: @tgamblin this concern is ostensibly resolved in https://github.com/spack/spack/pull/36347/commits/c92938fe6afa6725388dd3e4c4234b793602675b, but see https://github.com/spack/spack/pull/36347#issuecomment-1481973196 for a more-careful treatment of the related concerns

File: lib/spack/spack/solver/asp.py
Comment (scheibelp) [reply lib/spack/spack/solver/asp.py]: @haampie this concern is ostensibly resolved in https://github.com/spack/spack/commit/c92938fe6afa6725388dd3e4c4234b793602675b, but see https://github.com/spack/spack/pull/36347#issuecomment-1481973196 for a more-careful treatment of the related concerns

None of the problems addressed by this PR have to do with the construction of sets, so cannot be resolved by making `hash` and `==` consistent. I've asserted here that the semantics for `==` and `satisfies` are what they should be, and have seen no argument in this PR making a case for (or convincing me) otherwise, so have trouble understanding how "fixing" that is the solution (if it in fact would resolve this problem).

File: lib/spack/spack/solver/asp.py
Comment (scheibelp) [reply lib/spack/spack/solver/asp.py]: That is now done, so I will collapse this as resolved (the other comments sufficiently express the remaining concerns).

File: lib/spack/spack/solver/asp.py
Comment (scheibelp) [reply lib/spack/spack/solver/asp.py:1880]: `versions` can be a list, range, or a single version. Using e.g. `satisfies` would probably be sufficient for my needs in this PR, but I don't think it would be sufficient for the scenarios it was initially created for (avoid the concretizer picking `1.x.y` instead of `1` if there is an explicit `1` version in the package).

Admittedly nothing fails in `spec_semantics.py`, `spec_dag.py`, `concretize.py`, or `concretize_requirements.py` if I trade `==` for `.satisfies`, which I'd expect to be the modules to cover this change, but again I don't trust that is in fact sensible given what I interpret its intent to be. The commit which adds this at https://github.com/spack/spack/commit/ada2fa36a9aae3b2f328ac5c553c30cf25bd79df doesn't add the test and is part of the much larger https://github.com/spack/spack/pull/19501 (I tried all of the test modules which looked relevant from that PR though).

Comment (haampie) [reply lib/spack/spack/solver/asp.py:1880]: Okay, let's leave it in the questionable state it's in and I'll fix it in my pr.

File: lib/spack/spack/solver/asp.py
Comment (haampie) [reply lib/spack/spack/solver/asp.py]: Right, I'm saying the exact match logic could've been easier if one wouldn't have to worry about version == git version. Since it's just one instance now, and the long comment could be dropped, it's fine.

File: lib/spack/spack/version.py
Comment (haampie) [reply lib/spack/spack/version.py]: Can you still remove this?

File: lib/spack/spack/test/concretize_requirements.py
Comment (haampie) [reply lib/spack/spack/test/concretize_requirements.py:261]: Ping

**Review by haampie — CHANGES_REQUESTED** [review]
See above

**Review by haampie — APPROVED** [review]

**Review by tgamblin — COMMENTED** [review]
LGTM

File: lib/spack/spack/version.py
Comment (haampie) [inline lib/spack/spack/version.py:642]: This is very bug prone. If no version is imposed by the user, it's looked up and something is assigned. So you can have `v1 = abcdef=1.2.3` and `v2 = abcdef` for which `v1.satisfies(v2)` but `v2 < v1` (as it causes a lookup of `ref_version`).

I maintain that we should just remove this lookup nonsense, it's just introducing too many edge cases, it's too opaque, and generally useless.

Comment (scheibelp) [reply lib/spack/spack/version.py:642]: I'd rather have a centralized conversation on this (if there is such a conversation, can you link to it here?).

In this particular branch of the `if` statement, neither version is supplied by the user. So we are comparing things like `foo@abcdef...` and `foo@abcdef...`, they are equal if the git IDs are equal as presented by the user.

There *is* a bug if the lookups would render to the same git ID: (e.g. a commit and a tag) e.g. `foo@abcdef...` and `foo@git.tag1` (if `tag1` resolves to `abcdef...`), although that's separate from what you mention here.

File: lib/spack/spack/solver/asp.py
Comment (alalazo) [inline lib/spack/spack/solver/asp.py:2166]: @tgamblin @scheibelp Sorry if I came late to reviewing this, and previously stopped at reading the title. My understanding was that this PR allowed unknown git versions to be used in package requirements, which is :+1: . 

Reading the implementation (while resolving conflicts in other PRs) it seems to me that this PR makes any non-existing version in package requirements an allowed one. I think this is a bug. For example:
```yaml
packages:
  zlib:
    require: "@2.0" # This version doesn't exist and might be the result of a typo
```
With the previous behavior on develop we would get:
```console
$ spack spec zlib
==> Error: Cannot satisfy the requirements in packages.yaml for the 'zlib' package. You may want to delete them to proceed with concretization. To check where the requirements are defined run 'spack config blame packages'
```
which, despite the error message that can be improved, is what I expect by a requirement. This PR made it such that:
```console
$ spack spec zlib
Input spec
--------------------------------
zlib

Concretized
--------------------------------
zlib@2.0%gcc@9.4.0+optimize+pic+shared build_system=makefile arch=linux-ubuntu20.04-icelake
```
which in my opinion is wrong, since no version 2.0 exists for `zlib`. Was that on purpose? Shouldn't we restrict this logic to only git versions?

Comment (scheibelp) [reply lib/spack/spack/solver/asp.py:2166]: > This version doesn't exist and might be the result of a typo

> which in my opinion is wrong, since no version 2.0 exists for zlib

If the user says it must be `zlib@2.0` then I think that's what it should be: if you say `spack spec zlib@2.0` on the command line then that's what we get (and Spack tries to find 2.0 from zlib's download link).

If nothing else, right now if a user says `require: foo@hash=1.5` and `1.5` doesn't exist in the package.py, then we want Spack to still allow it. I think it's consistent to extend that to all versions mandated in the config.

Printing a warning message might be in order if the package can't be fetched, but otherwise I would say this is the desired behavior. If you want to prevent declaration of number versions that don't exist (e.g. when not associated with a git hash), then that would work for the use cases I created it for, but would be inconsistent (for the reasons outlined above).

Comment (alalazo) [reply lib/spack/spack/solver/asp.py:2166]: I think it's inconsistent somehow. If the user says `+shared` we don't allow a `shared` variant where it doesn't exist. I think the only case where popping a non-git version into existence make sense are externals, since it's software already installed - and somebody put the information on its version in `packages.yaml`.

Comment (alalazo) [reply lib/spack/spack/solver/asp.py:2166]: > If nothing else, right now if a user says require: foo@hash=1.5 and 1.5 doesn't exist in the package.py, then we want Spack to still allow it. 

That's a git version, so I'm not questioning that. It's really "regular" versions that shouldn't be added in my opinion to the ~allowed~ declared ones if they appear in requirements.

Comment (scheibelp) [reply lib/spack/spack/solver/asp.py:2166]: If I do

```
$ spack spec zlib@2.0
Input spec
--------------------------------
zlib@2.0

Concretized
--------------------------------
zlib@2.0%apple-clang@12.0.0+optimize+pic+shared build_system=makefile arch=darwin-bigsur-skylake
```

~do~ to make sure, are you also arguing for not doing that anymore?

Comment (haampie) [reply lib/spack/spack/solver/asp.py:2166]: The problem is ultimately the ambiguity of `@2.0` syntax. I can sort of get Peter's interpretation, but I must say I find it a bit far-fetched to turn a requirement into a version definition. At least to me it looks different from specifying a version on the command line.

In any case, with #36273 the ambiguity can be resolved since it introduces syntax to distinguish between a version range and a specific version.

Comment (alalazo) [reply lib/spack/spack/solver/asp.py:2166]: > are you also arguing for not doing that anymore?

No, or at least not necessarily My points:
- Command line might be used for quick tests, so we should allow something like this (maybe gated to an argument, but that would be a separate discussion)
- We moved an early error to a later stage, where it may potentially hang waiting for an answer
- To require a given version and be sure Spack doesn't invent a new one, I am now forced to use a range. This is a change in behavior on how the configuration is treated.

That made me think this was a bug that slipped through. Do you agree that a configuration that previously would have failed almost immediately, can now potentially hang in CI?

File: lib/spack/spack/solver/asp.py
Comment (alalazo) [inline lib/spack/spack/solver/asp.py:1640]: Can anybody remind me what is `defined_version`[^1] and explain why it has an hard-coded weight of 1? This is purely for my knowledge and to record the explanation somewhere.

[^1]: I assume it's the `=version` part of a git version, but want to double check.

File: lib/spack/spack/solver/asp.py
Comment (scheibelp) [reply lib/spack/spack/solver/asp.py:2166]: > We moved an early error to a later stage, where it may potentially hang waiting for an answer

To be clear, I don't think it's a spack error: it's a user error; if they demand a version then spack has to give it to them; if they demand the wrong version, then spack has to try (and fail). It shouldn't have failed where it was failing before because it couldn't know it was impossible at concretization time (in the same way it can't fail at concretization time when asking for `spack spec zlib@2.0`).

> To require a given version and be sure Spack doesn't invent a new one, I am now forced to use a range.

We could create a syntax for specifying that the version has to be defined in the package.py, but other than that the same issue applies to `spack spec zlib@2.0`. This may be what @haampie refers to as an ambiguity (although I thought that was referring to another set of issues regarding version comparisons)

File: lib/spack/spack/solver/asp.py
Comment (scheibelp) [reply lib/spack/spack/solver/asp.py:1640]: If you have `hash=number` then `defined_version` is the `number`. It has a hard-coded weight of 1 because it is the next possible weight after 0 (i.e. we add either 1 or 2 items in this "dimension" the first having weight 0, and the next one - if added - has weight 1).

File: lib/spack/spack/solver/asp.py
Comment (haampie) [reply lib/spack/spack/solver/asp.py:2166]: I agree with @alalazo it's much more likely the user makes a typo than that they intentionally request a non-existing version (and happen to know of this curious behavior of Spack). Personally I never use non-existing versions, and it's not _that_ hard to just "register" the version if you need something not there yet.

Regarding ambiguity I mean: you can interpret it as "use version 2.0.something" or "use 2.0 exactly". If you're reading it as "use 2.0.something" you would expect an error if there is no 2.0.something. If you read it as 2.0 exactly you would _maybe_ expect spack to make that version up (but apparently both @alalazo and me did not expect that to happen).

Comment (scheibelp) [reply lib/spack/spack/solver/asp.py:2166]: > Personally I never use non-existing versions, and it's not that hard to just "register" the version if you need something not there yet.

Firstly by "register" I think you mean "add the version to package.py". Enough users that I interact with need this (i.e. the ability to request versions *not* in the package.py) enough that I consider it necessary; although for the users I'm thinking of, they just need `require: @hash=number` to do this, rather than `require: @number`.

I could see the argument that it's unlikely for users to want a numbered version not in the package.py, but like I said we already (i.e. before this PR) allowed Spack to do this in other circumstances (the command line, or when specified as a preference). 

Are you (@haampie and @alalazo) proposing:

* Forbid `spack spec zlib@2.0` (currently allowed)
* Forbid `packages:zlib:version:[2.0]` (currently allowed)
* Allow `spack spec zlib@hash=2.0` (pretending zlib has a git repo)
* Allow `packages:zlib:verssion:[hassh=2.0]`

?

File: lib/spack/spack/solver/asp.py
Comment (alalazo) [reply lib/spack/spack/solver/asp.py:1640]: I don't think this will cause issues right now, but note that this introduces multiple versions with weight 1 in a context where there are multiple versions of weight 0 (and one that is very likely enforced, which is the one above). So, we'll likely have:
```
0 -> this version
0 -> the default preferred version 
1 -> the defined_version
1 -> the next default preferred version
2 -> ...
```

File: lib/spack/spack/solver/asp.py
Comment (alalazo) [reply lib/spack/spack/solver/asp.py:2166]: I would be fine already with:
- Allow `@<any git version>` in requirements to define the version
- Get back the previous behavior for non-git versions

and discuss the other points later. Right now any requirement on version is somehow self-fulfilling if the version doesn't exist (and bound to fail later). 

---

That said, I think defining unknown versions make sense only for `git versions` or already installed software (buildcache, local DB, or externals). We can probably allow:
```
packages:zlib:version:[2.0]
```
But that shouldn't be defining a version 2.0 if it doesn't exist somewhere else. So, clingo would prefer to use 2.0 but if it isn't defined somewhere else, would use the next best.

Comment (Paul-Ferrell) [reply lib/spack/spack/solver/asp.py:2166]: When we define an environment, we also validate that the concretization output matches all of the versions we specify. So that should, in theory, catch the fact that the defined default in packages doesn't match what we get. 

A warning would be nice though, that way we see the discrepancy faster. 

Also, having packages define default versions that are undefined in the package causes concretization errors. The version doesn't end up in the concretization facts, version checks in `when` clauses all fail for that version.

Comment (scheibelp) [reply lib/spack/spack/solver/asp.py:2166]: @alalazo

> But that shouldn't be defining a version 2.0 if it doesn't exist somewhere else.

To be clear, that was the behavior in Spack prior to this PR: putting versions in the preferences defines them as possible versions (i.e. if `2.0` is a version preference, `spack spec zlib` will give you `zlib@2.0`). ~Not to say you can't undo that~ You can undo that, but it will take (slightly more) than reverting parts of this PR to get what you want.

Therefore

> Get back the previous behavior for non-git versions

the "previous behavior" doesn't seem to be a valid given your own stated desires. I suggest outlining the things you want to be true (e.g. as done in https://github.com/spack/spack/pull/36347#discussion_r1153902072) and then we can make them true (ignoring however they were behaving in the past).

Comment (alalazo) [reply lib/spack/spack/solver/asp.py:2166]: > To be clear, that was the behavior in Spack prior to this PR [ .. ]

I know, sorry if the reply was confusing but:
1. My immediate worry is if we can get back the old behavior for non-git versions in requirements
2. Then you asked what I thought about the other parts (preferences, CLI) so I gave an answer to that too - but changes there can definitely wait

Also, see this issue we got today https://github.com/spack/spack/issues/36574 which is related to the discussion.

Comment (scheibelp) [reply lib/spack/spack/solver/asp.py:2166]: It is related but FWIW if you check out 9a93f223d6 (the commit before this was merged) you can still reproduce https://github.com/spack/spack/issues/36574. In some sense, my goal will be to extend this to do the opposite of what was decided there (for `hash=number` versions at least).

Preventing users from introducing new non-hash versions in preferences will not conflict with this PR or the users I know of that would benefit from it, so I only oppose it on account of it being inconsistent (I am willing to be overridden on that).

Comment (scheibelp) [reply lib/spack/spack/solver/asp.py:2166]: > It is related 

I said this because I assumed that package requirements would have the same problem, but they don't: https://github.com/spack/spack/pull/36589; therefore that issue is not related to this PR (or at least not caused by it).

**Pull request merged**
